src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -73099,6 +73099,335 @@ static void Main()
+            Assert.False(c2.IsDefinition);
+
+            AssertHashCodesMatch(cDefinition, c2);
+
+            var i2 = c2.GetTypeMember("I");
+            Assert.Equal("C<T!>.I", i2.ToTestDisplayString(includeNonNullable: true));
+            Assert.False(i2.IsDefinition);
+
+            AssertHashCodesMatch(iDefinition, i2);
+
+            var c3 = cDefinition.Construct(ImmutableArray.Create(TypeWithAnnotations.Create(cDefinition.TypeParameters.Single(), NullableAnnotation.Annotated)));
+            Assert.Equal("C<T?>", c3.ToTestDisplayString(includeNonNullable: true));
+            Assert.False(c3.IsDefinition);
+
+            AssertHashCodesMatch(cDefinition, c3);
+
+            var i3 = c3.GetTypeMember("I");
+            Assert.Equal("C<T?>.I", i3.ToTestDisplayString(includeNonNullable: true));
+            Assert.False(i3.IsDefinition);
+
+            AssertHashCodesMatch(iDefinition, i3);
+        }
+
+        [Fact]
+        [WorkItem(30673, "https://github.com/dotnet/roslyn/issues/30673")]
+        public void TypeSymbolGetHashCode_ContainingType_GenericNestedType()
+        {
+            var text = @"
+class C<T> where T : class
+{
+    interface I<U> where U : class { }
+}
+";
+
+            var comp = CreateNullableCompilation(text);
+            var iDefinition = comp.GetMember<NamedTypeSymbol>("C.I");
+            Assert.Equal("C<T>.I<U>", iDefinition.ToTestDisplayString(includeNonNullable: true));
+            Assert.True(iDefinition.IsDefinition);
+
+            // Construct from iDefinition with annotated U from iDefinition
+            var i1 = iDefinition.Construct(ImmutableArray.Create(TypeWithAnnotations.Create(iDefinition.TypeParameters.Single(), NullableAnnotation.Annotated)));
+            Assert.Equal("C<T>.I<U?>", i1.ToTestDisplayString(includeNonNullable: true));
+            AssertHashCodesMatch(iDefinition, i1);
+
+            var cDefinition = iDefinition.ContainingType;
+            Assert.Equal("C<T>", cDefinition.ToTestDisplayString(includeNonNullable: true));
+            Assert.True(cDefinition.IsDefinition);
+
+            // Construct from cDefinition with unannotated T from cDefinition
+            var c2 = cDefinition.Construct(ImmutableArray.Create(TypeWithAnnotations.Create(cDefinition.TypeParameters.Single(), NullableAnnotation.NotAnnotated)));
+            var i2 = c2.GetTypeMember("I");
+            Assert.Equal("C<T!>.I<U>", i2.ToTestDisplayString(includeNonNullable: true));
+            Assert.Same(i2.OriginalDefinition, iDefinition);
+            AssertHashCodesMatch(i2, iDefinition);
+
+            // Construct from i2 with U from iDefinition
+            var i2a = i2.Construct(iDefinition.TypeParameters.Single());
+            Assert.Equal("C<T!>.I<U>", i2a.ToTestDisplayString(includeNonNullable: true));
+            AssertHashCodesMatch(iDefinition, i2a);
+
+            // Construct from i2 with annotated U from iDefinition
```

Comment:
>Construct from i2 with annotated U from iDefinition [](start = 15, length = 51)

:memo: I couldn't do this in VB. It hits an assertion that we're dealing with a definition, in `NamedTypeSymbol.Construct`. Is that a bug?

```vb
            ' Construct from i2 with modified U from iDefinition
            Dim i2b = i2.Construct(TypeSubstitution.Create(iDefinition, iDefinition.TypeParameters, modifiedU))

```